src/core/algorithms/dd/dd.h

```diff
@@ -1,22 +1,96 @@
 #pragma once
 
+#include <compare>
 #include <list>
+#include <optional>
 #include <sstream>
 #include <string>
+#include <utility>
 #include <vector>
 
+#include <boost/test/tools/floating_point_comparison.hpp>
+
+#include "model/types/builtin.h"
+#include "model/types/double_type.h"
+
 namespace model {
 
+static double constexpr kRelativeTolerance =
+        DoubleType::kDefaultEpsCount * std::numeric_limits<model::Double>::epsilon();
+
+static bool IsEqual(double first_value, double second_value) {
+    return boost::math::fpc::close_at_tolerance<model::Double>(kRelativeTolerance)(first_value,
+                                                                                   second_value);
+}
+
+static bool Less(double first_value, double second_value) {
+    return first_value < second_value && !IsEqual(first_value, second_value);
+}
+
+static bool Greater(double first_value, double second_value) {
+    return first_value > second_value && !IsEqual(first_value, second_value);
+}
+
+static bool LessOrEqual(double first_value, double second_value) {
+    return !Greater(first_value, second_value);
+}
+
+static bool GreaterOrEqual(double first_value, double second_value) {
+    return !Less(first_value, second_value);
+}
+
 struct DFConstraint {
     double lower_bound;
     double upper_bound;
 
+    auto operator<=>(DFConstraint const& other) const {
+        if (!IsEqual(lower_bound, other.lower_bound)) {
+            return Less(lower_bound, other.lower_bound) ? std::strong_ordering::less
+                                                        : std::strong_ordering::greater;
+        }
+        if (!IsEqual(upper_bound, other.upper_bound)) {
+            return Less(upper_bound, other.upper_bound) ? std::strong_ordering::less
+                                                        : std::strong_ordering::greater;
+        }
+        return std::strong_ordering::equal;
+    }
+
     bool operator==(DFConstraint const& other) const {
-        return lower_bound == other.lower_bound && upper_bound == other.upper_bound;
+        return IsEqual(lower_bound, other.lower_bound) && IsEqual(upper_bound, other.upper_bound);
+    }
```

Comment:
This code doesn't compile without `operator==`. As far as I can see, only defaulted `operator<=>` can implicitly define `operator==`. Otherwise, both of the primary operators should be defined: https://brevzin.github.io/c++/2019/07/28/comparisons-cpp20/#new-operator-abilities

Reply:
Oh, okay, didn't know that. Thanks!